File: 0210-course-schedule-ii/0210-course-schedule-ii.java
class Solution {
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        ArrayList<ArrayList<Integer>> adj = new ArrayList<>();
        for(int i=0; i<numCourses; i++){
            adj.add(new ArrayList<>());
        }

        for(int[] prereq: prerequisites){
            adj.get(prereq[1]).add(prereq[0]);
        }

        int[] indegree = new int[numCourses];

        for(ArrayList<Integer> adjacent: adj){
            for(int i: adjacent){
                indegree[i]++;
            }
        }

        Queue<Integer> q = new LinkedList<>();

        for(int i=0; i<numCourses; i++){
            if(indegree[i]==0){
                q.add(i);
            }
        }

        ArrayList<Integer> res = new ArrayList<>();

        while(!q.isEmpty()){
            int node = q.poll();
            res.add(node);

            for(int neigh: adj.get(node)){
                indegree[neigh]--;
                if(indegree[neigh]==0){
                    q.add(neigh);
                }
            }
        }

        if(res.size()!=numCourses){
            return new int[0];
        }

        int[] result = new int[numCourses];

        for(int i=0; i<numCourses; i++){
            result[i] = res.get(i);
        }

        return result;
    }
}

File: 0210-course-schedule-ii/0210-course-schedule-ii.py
class Solution(object):
    def findOrder(self, numCourses, prerequisites):
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype: List[int]
        """
        def dfs(node, path_visited):
            visited[node] = True
            path_visited[node] = True

            for next in adj[node]:
                if not visited[next]:
                    if dfs(next, path_visited):
                        return True
                elif path_visited[next]:
                    return True

            path_visited[node] = False
            res.append(node)
            return False
        
        #Adj 
        adj = {}

        for i in range(numCourses):
            adj[i] = []
        
        for v, u in prerequisites:
            adj[u].append(v)
        
        visited = [False] * numCourses
        path_visited = [False] * numCourses
        res = []

        for i in range(numCourses):
            if not visited[i]:
                if dfs(i, path_visited):
                    return []

        return res[::-1]